usersession/userd/launcher.go

```diff
@@ -170,6 +185,216 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface. The desktopFileID is described here:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+func (s *Launcher) OpenDesktopEntryEnv(desktopFileID string, env []string, sender dbus.Sender) *dbus.Error {
+	desktopFile, err := desktopFileIDToFilename(existsOnFileSystem, desktopFileID)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	err = verifyDesktopFileLocation(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	exec_command, err := readExecCommandFromDesktopFile(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	args, err := parseExecCommand(exec_command)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		if !strutil.ListContains(allowedEnvVars, strings.SplitN(e, "=", 2)[0]) {
+			return dbus.MakeFailedError(fmt.Errorf("Supplied environment variable %q is not allowed", e))
+		}
+
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	// XXX: this avoids defunct processes but causes userd to persist
+	// until all children are gone (currently, this is not a problem since
+	// userd is long running once started)
```

Comment:
Are you sure this comment is correct?  Go doesn't wait for all goroutines to exit before the process dies: you'd need to have some kind of explicit serialisation for this to be true.

I would think userd exiting while the app is still running should just result in the app being reparented to pid 1, which is what we'd want.

Reply:
The text originates with @jdstrand: https://github.com/snapcore/snapd/pull/8699#discussion_r432845187 and I had no reason to disbelieve it. (I'm not sufficiently familiar with go know otherwise.)

Reply:
Please note that I was hoping others would confirm the approach and advise on a proper comment string. Please feel free to adjust as necessary.

Reply:
I think the bigger concern is whether this goroutine represents another OS thread.  From the look of it, the goroutine will be blocked in a `waitid` syscall, so the OS thread can't be reused by another goroutine.  So a snap with `desktop-launch` can cause userd to spawn an arbitrary number of threads.

Ideally we could dissociate from the child process as soon as it has been successfully forked, rather than it being waitable.

Reply:
> I think the bigger concern is whether this goroutine represents another OS thread. From the look of it, the goroutine will be blocked in a `waitid` syscall, so the OS thread can't be reused by another goroutine. So a snap with `desktop-launch` can cause userd to spawn an arbitrary number of threads.
> 
> Ideally we could dissociate from the child process as soon as it has been successfully forked, rather than it being waitable.

This was discussed in more depth here: https://github.com/snapcore/snapd/pull/8699#discussion_r432845187. As mentioned there, the current approach was better than calling out to 'sh' but there may be better options.

Reply:
The traditional way to dissociate yourself from a child process is to do a double fork and have the first child process exit, so the process that calls exec() is a grandchild.  That's tricky to get right in a multi-threaded environment, which is why glib these days uses a `gio-launch-desktop` helper as the intermediate child these days.  That is similar to the previous shell approach, but doesn't involve a shell trying to reinterpret the arguments.

It might still be possible to safely use the shell safely as an intermediate process, something like this:

    cmd := exec.Command("/bin/sh", "-c", `"$0" "$@" &`, args...)

This should return immediately, so calling `cmd.Run()` should be safe, and we don't need to re-serialise the command to be shell safe.

Reply:
Right, go isn't doing as any favors for the traditional methods.

If people don't care for the current technique, I'm pretty sure I'd prefer a helper over reintroducing the shell since we can completely control the helper and wouldn't have to worry about what shell /bin/sh was, etc.

Reply:
I talked with @pedronis about this yesterday, and he's not a fan of using `/bin/sh` as an intermediate process either.  Instead, he was more in favour of using a helper executable to act as the intermediate.

If we continue doing everything in Go, this helper could be a `snap` sub-commnand (maybe `snap routine desktop-launch-helper`).  Depending on what the scope for this interface is, it could also be something written in C: if Ubuntu Core is the only target, then a helper linked against glib/gio and shipped with the `core18` or `core20` snap is a possibility.

If we have a dedicated helper, we've also got a bit more flexibility about where the desktop file parsing occurs.  While we could just pass it an argument vector to execute, we could also pass it the desktop file name or desktop ID, and have the parsing occur there. That could be beneficial if we can use glib in the helper, since we could rely on its well tested desktop file parsing code.

Reply:
Switched to `systemd-run` as the proxy,